com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/ProtocolServer.java

```diff
@@ -33,42 +33,36 @@
      */
     public ProtocolServer(InputStream input, OutputStream output, IProviderContext context) {
         super(input, output);
-        this.debugAdapter = new DebugAdapter((debugEvent, willSendLater) -> {
-            // If the protocolServer has been stopped, it'll no longer receive any event.
-            if (!terminateSession) {
-                if (willSendLater) {
-                    sendEventLater(debugEvent.type, debugEvent);
-                } else {
-                    sendEvent(debugEvent.type, debugEvent);
-                }
-            }
+        this.debugAdapter = new DebugAdapter((message) -> {
+            sendMessage(message);
         }, context);
     }
 
     /**
      * A while-loop to parse input data and send output data constantly.
      */
-    public void start() {
+    public void run() {
         usageDataSession.reportStart();
-        super.start();
-    }
-
-    /**
-     * Sets terminateSession flag to true. And the dispatcher loop will be terminated after current dispatching operation finishes.
-     */
-    public void stop() {
+        super.run();
         usageDataSession.reportStop();
-        super.stop();
         usageDataSession.submitUsageData();
     }
 
-    protected void dispatchRequest(Messages.Request request) {
-        Messages.Response response = this.debugAdapter.dispatchRequest(request);
-        if (request.command.equals("disconnect")) {
-            stop();
+    @Override
+    protected void sendMessage(Messages.ProtocolMessage message) {
+        super.sendMessage(message);
+        if (message instanceof Messages.Response) {
+            usageDataSession.recordResponse((Messages.Response) message);
+        } else if (message instanceof Messages.Request) {
+            usageDataSession.recordRequest((Messages.Request) message);
         }
-        sendMessage(response);
-        usageDataSession.recordResponse(response);
+    }
+
+    protected void dispatchRequest(Messages.Request request) {
+        usageDataSession.recordRequest(request);
+        this.debugAdapter.dispatchRequest(request).thenAccept((response) -> {
+            sendMessage(response);
+        });
```

Comment:
We need to have some code to handle exception from the promise. A promise isn't always completed. It can fail with an exception too.

Reply:
fixed.
Use `future.whenComplete()` to handle normal response and exceptional response.